bench/irmin-pack/tree.ml

```diff
@@ -0,0 +1,398 @@
+open Lwt.Infix
+open Bench_common
+
+type key = string list [@@deriving yojson]
+type hash = string [@@deriving yojson]
+type message = string [@@deriving yojson]
+
+type op =
+  | Add of key * string
+  | Remove of key
+  | Find of key * bool
+  | Mem of key * bool
+  | Mem_tree of key * bool
+  | Commit of hash * int64 * message * hash list
+  | Checkout of hash
+  | Copy of key * key
+[@@deriving yojson]
+
+let commits = ref [||]
+
+module Parse_log = struct
+  let prepare_log file =
+    if Sys.file_exists file then (
+      let cmd = Filename.quote_command "tar" [ "-zxvf"; file ] in
+      Fmt.epr "exec %s\n %!" cmd;
+      match Sys.command cmd with
+      | 0 -> ()
+      | n ->
+          Fmt.failwith
+            "Failed to set up the test environment: command `%s' exited with \
+             non-zero exit code %d"
+            cmd n)
+    else Fmt.failwith "File %s not found" file
+
+  let operation_list = ref []
+
+  let read_commits file commits ncommits =
+    let json = Yojson.Safe.from_file file in
+    let operations = Yojson.Safe.Util.to_list json in
+    let rec aux index_op index_commit ls =
+      match ls with
+      | [] ->
+          if index_commit < ncommits then
+            Fmt.epr
+              "Only %d commits available in the log file, proceeding...\n%!"
+              index_commit;
+          index_commit
+      | _ when index_commit >= ncommits -> index_commit
+      | json :: tl -> (
+          match op_of_yojson json with
+          | Ok (Commit _ as x) ->
+              operation_list := x :: !operation_list;
+              commits.(index_commit) <- List.rev !operation_list;
+              operation_list := [];
+              (aux [@tailcall]) (index_op + 1) (index_commit + 1) tl
+          | Ok x ->
+              operation_list := x :: !operation_list;
+              (aux [@tailcall]) (index_op + 1) index_commit tl
+          | Error s -> Fmt.failwith "error op_of_yosjon %s\n%!" s)
+    in
+    aux 0 0 operations
+
+  let populate_array log ncommits =
+    let commits_tmp = Array.init ncommits (fun _ -> []) in
+    let total = read_commits log commits_tmp ncommits in
+    commits := commits_tmp;
+    total
+
+  let cleanup () = commits := [||]
+end
+
+module Generate_trees_from_log
+    (Store : Irmin.S with type contents = string and type key = string list) =
+struct
+  type t = {
+    mutable tree : Store.tree;
+    mutable prev_commit : Store.hash option;
+  }
```

Comment:
I wonder if it's simpler to `Lwt_list.fold_left_s` over this type as an accumulator rather than mutating it on each iteration of an `Lwt_list.iteri_s`. `caml_modify` can be a bit slow, so it might be faster in addition.